Add P95 and P99 as standard aggregations

uhoh-itsmaciek opened the pull request:
I'm running with these in our environment, and they seem to be working okay. The approach for calculating these values is pretty naive, but a quick look didn't suggest anything better for this use case.

j-white commented on api/src/main/java/org/opennms/newts/api/query/StandardAggregationFunctions.java:
```diff
@@ -69,6 +71,45 @@ public Double apply(Collection<Double> input) {
+    P95 {
+
+        @Override
+        public Double apply(Collection<Double> input) {
```
We should aggregate these function and pass the percentile as an argument instead of defining it twice.

uhoh-itsmaciek replied:
Good call, will do.

j-white commented:
Also, can we add tests that verify the percentile calculations. The current algorithm doesn't look right to me.

uhoh-itsmaciek commented:
Sure, I'll add tests.

uhoh-itsmaciek commented:
Added tests (added some basic test for the existing MIN, MAX, and AVERAGE since there were none; let me know if you'd like to move that out to a separate PR) and refactored to use a helper function. My Java's really really rusty, so please let me know if I'm doing something really dumb.

j-white commented:
Thanks @uhoh-itsmaciek, I think we're in good shape now.

We'll need you to sign an OCA before I can merge this though. Details at  http://www.opennms.org/documentation/ContributorAgreement.pdf

I added a CONTRIBUTING.md file to help make this clearer in the future.

fooker commented on api/src/main/java/org/opennms/newts/api/query/StandardAggregationFunctions.java:
```diff
@@ -69,6 +71,37 @@ public Double apply(Collection<Double> input) {
+            }
+        }
+        if (sorted.isEmpty()) {
+            return Double.MAX_VALUE;
```
I'm wondering if MAX_VALUE is the right thing here? Other aggregations use NaN in such a case.

j-white replied:
Nice catch.

uhoh-itsmaciek replied:
Well, MIN retuns `Double.MAX_VALUE`, MAX returns `Double.MIN_VALUE`, and AVERAGE returns 0. But yeah, `Double.NaN` probably makes sense here.

uhoh-itsmaciek commented:
Great, thanks. The CLA may take me a while, unfortunately; I'll follow up.

uhoh-itsmaciek commented:
Sorry, I dropped the ball on the CLA. We're no longer using newts (read-time roll-ups were not the right mechanism for us), so I doubt I'll be finish this process--sorry about that. If you institute a simpler semi-automated electronic CLA mechanism (like, say, in https://github.com/discourse/discourse/pull/4504#issuecomment-254055944 ), I'm happy to go through that, but in the meantime, I'll close the PR.
